src/compiler/types.ts

```diff
@@ -3609,7 +3609,13 @@ namespace ts {
          * This is intended to be the first top-level import/export,
          * but could be arbitrarily nested (e.g. `import.meta`).
          */
-        /* @internal */ externalModuleIndicator?: Node;
+        /* @internal */ externalModuleIndicator?: Node | true;
+        /**
+         * The callback used to set the external module indicator - this is saved to
+         * be later reused during incremental reparsing, which otherwise lacks the information
+         * to set this field
+         */
+        /* @internal */ setExternalModuleIndicator?: (file: SourceFile) => void;

```

Comment:
Because the module indicator now depends on the compiler options the source file was created with, which the source file retains no direct reference to, and small incremental edits (the kind that don't do a full source file rebuild) don't pass in full options, either, they just reuse the old source file's data - so we have to keep the callback around to reuse in that incremental scenario.

Reply:
Hm, I forgot that parsing in general, much less incremental parsing, doesn’t have access to CompilerOptions; only the (previously very small) subset of them that it needs. I’m a little worried that this will become a footgun whereby we or other API consumers will accidentally close over and retain something they shouldn’t in this function (it you haven’t seen the implementation, you don’t know a reference to it is stored on the SourceFile—the property is even erased from the public API!). What if `CreateSourceFileOptions` took the `moduleDetection`, `module`, and `target` from CompilerOptions and stored those, such that the function currently returned by `getSetExternalModuleIndicator` would have everything it needs to run on the SourceFile after creation? It would be more properties to store, but more straightforward. It took me a minute to understand the purpose of this indirection.

Reply:
Hm, while that would be enough for recalculating the correct closure, it doesn't do much to help the `impliedNodeFormat` field calculation, which needs extensive host  availability anyway. I'm disinclined to make these look like `CompilerOptions` because of that - if you just passed in compiler options, you'd be broken in any nodenext type scenarios.

Reply:
`impliedNodeFormat` now gets passed into `createSourceFile` (and I wasn’t proposing removing that, for the reason you mentioned), and I was assuming it can’t change during `updateSourceFile` (incremental parsing)—is that right? If so, then incremental parsing doesn’t need to worry about `impliedNodeFormat`.